core/src/main/java/net/opentsdb/query/hacluster/HAClusterFactory.java

```diff
@@ -173,417 +124,135 @@ public void setupGraph(final QueryPipelineContext context,
-      builder = cluster_config.toBuilder();
-      builder.setHasBeenSetup(true);
-
-      if (Strings.isNullOrEmpty(cluster_config.getMergeAggregator())) {
-        builder.setMergeAggregator(tsdb.getConfig().getString(
-            getConfigKey(AGGREGATOR_KEY)));
-      }
-      if (Strings.isNullOrEmpty(cluster_config.getPrimaryTimeout())) {
-        builder.setPrimaryTimeout(tsdb.getConfig().getString(
-            getConfigKey(PRIMARY_KEY)));
-      }
-      if (Strings.isNullOrEmpty(cluster_config.getSecondaryTimeout())) {
-        builder.setSecondaryTimeout(tsdb.getConfig().getString(
-            getConfigKey(SECONDARY_KEY)));
-      }
-      
-      if (cluster_config.getDataSources().isEmpty() && 
-          cluster_config.getDataSourceConfigs().isEmpty()) {
-        // sub in the defaults.
-        synchronized (default_sources) {
-          builder.setDataSources(Lists.newArrayList(default_sources));
+      final List<String> dataSources = ((HAClusterConfig) config).getDataSources();
+      if (dataSources != null && !dataSources.isEmpty()) {
+        final List<TimeRouterConfigEntry> newSources =
+                Lists.newArrayListWithExpectedSize(dataSources.size());
+        for (TimeRouterConfigEntry entry : sources) {
+          if (dataSources.contains(entry.getSourceId())) {
+            newSources.add(entry);
+          }
         }
+        sources = newSources;
       }
-    } else {
-      builder = HAClusterConfig.newBuilder();
-      HAClusterConfig.cloneBuilder(config, builder);
-      builder.setMergeAggregator(tsdb.getConfig().getString(
-               getConfigKey(AGGREGATOR_KEY)))
-             .setPrimaryTimeout(tsdb.getConfig().getString(
-               getConfigKey(PRIMARY_KEY)))
-             .setSecondaryTimeout(tsdb.getConfig().getString(
-                getConfigKey(SECONDARY_KEY)))
-             .setHasBeenSetup(true);
-      synchronized (default_sources) {
-        builder.setDataSources(Lists.newArrayList(default_sources));
-      }
-    }
-
-    final String new_id = "ha_" + config.getId();
-    if (context.query().isTraceEnabled()) {
-      context.queryContext().logTrace("Elligible sources: " + builder.dataSources());
     }
 
-    // if there is only one source, drop the merger and ha config nodes
-    if (builder.dataSources().size() + builder.dataSourceConfigs().size() == 1) {
-      if (!builder.dataSources().isEmpty()) {
-        if (planner.context().tsdb().getRegistry().getPlugin(
-              TimeSeriesDataSourceFactory.class, 
-              builder.dataSources().get(0)) == null) {
-          throw new IllegalArgumentException("No data source found for: " 
-            + builder.dataSources().get(0));
+    List<TimeRouterConfigEntry> supportedSources = null;
+    for (int i = 0; i < sources.size(); i++) {
+      final TimeRouterConfigEntry entry = sources.get(i);
+      if (!entry.factory().supportsQuery(context, config)) {
```

Comment:
Should the removal of the non-supported sources happen in the Query planner as opposed to here ?

Reply:
Not in this case cause it could be that one HA source is valid but the other isn't. The planner doesn't know if the other source is HA or split or what so it's a bit better here I think.